### ISSUE TITLE
Flow types

### ISSUE THREAD

**Issue by steida** [redux-observable/redux-observable#258]
Does anybody is working on that?

**Comment by jayphelps** [comment]
Not as far as I know. If you make solid ones, we'd definitely love to have them included! ❤️  One important thing is that to provide any meaningful type definition for redux-observable you actually would also need a type definition for rxjs. I see there are some here https://github.com/flowtype/flow-typed/tree/master/definitions/npm/rxjs_v5.0.x but I haven't personally tried them.

**Comment by steida** [comment]
Why? Flow types can be spread aka ...

**Comment by jayphelps** [comment]
@steida sorry, I don't understand the question?

**Comment by steida** [comment]
Is it sure that rxjs flow types can not be reused & extended?

**Comment by jayphelps** [comment]
@steida you do not have to _duplicate_ rxjs typings. You just need them to exist and work correctly, which redux-observable's typings would build on top of. the ones in flow-typed probably will work fine, but I was just warning that I have not tested them and so they could have holes or incorrect types for some things because rxjs is a large project and full of complex type scenarios.

**Comment by steida** [comment]
@jayphelps rxjs flow types works well, I am using it

// custom epic type
https://github.com/este/este/blob/a2d28238a1ba34c364343b4779d4a5e818908d31/types/index.js#L105

// type refinements
https://github.com/este/este/blob/a2d28238a1ba34c364343b4779d4a5e818908d31/lib/users/epics.js#L21

**Comment by rafaelkallis** [comment]
@steida I would be willing to help. Currently working on a rxjs based framework that extends the Observable just like `ActionsObservable` does and I'm using flow. Shouldn't be hard, given we already have the typescript typings. I also do suggest to make the flow typings a **1 to 1** replica of the typescript typings. I don't see any syntax on the `redux-observable` TS typings that can't be ported to flow typings.

**Comment by jayphelps** [comment]
Our typings are likely about to change in #250 to better refine the dependency injection type.

**Comment by steida** [comment]
@jayphelps I believe DI should include getState as well, all in one argument. I had to workaround current approach, and it works perfectly. https://github.com/este/este/blob/a2d28238a1ba34c364343b4779d4a5e818908d31/lib/root-epic.js#L16

**Comment by jayphelps** [comment]
@steida can you clarify what you mean? the store and the DI are different arguments.

**Comment by steida** [comment]
Nothing more that's it's just practical. Look https://github.com/este/este/blob/a2d28238a1ba34c364343b4779d4a5e818908d31/lib/users/epics.js#L17 

I can add getState, whatever. It's typed. I don't know why getState and other dependencies should be passed as the third argument. 

![screen shot 2017-06-09 at 23 46 02](https://user-images.githubusercontent.com/66249/26995588-db4f8bfa-4d6d-11e7-9918-f58cc0c3f789.png)

**Comment by jayphelps** [comment]
@steida if you have a solid proposal, feel free to create a separate issue ticket RFC around it.

**Comment by steida** [comment]
Just create root epic and inject whatever in. It's more design pattern than lib feature.

```js
// @flow
import 'rxjs';
import type { Action, Dependencies, State } from '../types';
import type { Observable } from 'rxjs';
import { combineEpics } from 'redux-observable';
import { values } from 'ramda';

import * as usersEpics from './users/epics';

const importedEpics = {
  ...usersEpics,
};

const epics = values(importedEpics).filter(item => typeof item === 'function');

const rootEpic = (
  action$: Observable<Action>,
  { getState }: { getState: () => State },
  dependencies: Dependencies
) => combineEpics(...epics)(action$, { ...dependencies, getState });

export default rootEpic;
```

**Comment by rafaelkallis** [comment]
@steida I'm posting this here since there's no open issue yet, but it's not the right place to discuss.

Am I missing something or isn't `getState` already included, independent from DI?

Reference:

[createEpicMiddleware.js](https://github.com/redux-observable/redux-observable/blob/6384c9f22df10621127dbf7002023faa0935573c/src/createEpicMiddleware.js#L38-L39)
```js
const output$ = ('dependencies' in options)
  ? epic(action$, store, options.dependencies)
  : epic(action$, store);
```
---
[index.d.ts](https://github.com/redux-observable/redux-observable/blob/6384c9f22df10621127dbf7002023faa0935573c/index.d.ts#L32)
```ts
export declare interface Epic<T, S> {
  (action$: ActionsObservable<T>, store: MiddlewareAPI<S>): Observable<T>;
}
```

where `store: MiddlewareAPI<S>` has `getState`.

**Comment by steida** [comment]
I will create separate RFC later. The point is, not having two arguments when we can have only one.

**Comment by steida** [comment]
Yep! https://github.com/flowtype/flow-typed/tree/master/definitions/npm/redux-observable_v0.14.x

**Comment by steida** [comment]
Actually, it's broken. Better to not use it. https://github.com/flowtype/flow-typed/issues/1020

**Comment by wldcordeiro** [comment]
@steida it also doesn't spread in the original RxJS types so you get lack of coverage when you start using operators. 😞
